### ISSUE TITLE
AutomationElement for MenuItems has changed in a breaking way

### ISSUE THREAD

**Issue by lovettchris** [dotnet/winforms#10109]
### .NET version

.NET SDK:
 Version:   8.0.100-rc.1.23463.5
 Commit:    e7f4de8816

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.0-rc.1.23419.4
  Architecture: x64
  Commit:       92959931a3
  RID:          win-x64

.NET SDKs installed:
  3.1.426 [C:\Program Files\dotnet\sdk]
  6.0.415 [C:\Program Files\dotnet\sdk]
  7.0.402 [C:\Program Files\dotnet\sdk]
  8.0.100-rc.1.23463.5 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-rc.1.23421.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rc.1.23419.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.23 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0-rc.1.23420.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

I'm using .NET 7.0.  Same problem in .NET Framework 4.8 but not in .NET Framework 4.7.2

### Issue description

AutomationElements for menu items have disappeared so I cannot automate my application any more.  The app is https://github.com/microsoft/XmlNotepad and the unit tests all use AutomationElement and the tests work fine on commit c47d66e5ae1552d8a9e1b898b235a01f4b1f821c but not in the latest master branch.

### Steps to reproduce

1. Create new WinForms app with a menuStrip with one item "File" and put menu items in this File menu named "New" and "Open"
2. run the app
3. run C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x64\inspect.exe

Hover over the file menu, notice it has a little + sign next to it, but when you click the + no sub menu items appear.

This was not the case in .NET 4.7.2 where the sub menu items were all perfectly accessible.

Alternate repro:
1. clone https://github.com/microsoft/XmlNotepad,
2. load src\XmlNotepad.sln
3. Run all unit tests

They all fail on .NET 4.8, but they pass on .NET 4.7.2.  In particular you will see a difference in the menu item iteration in the ReloadMenuItems function where the expression `FindAll(TreeScope.Descendants,  new PropertyCondition(AutomationElement.ControlTypeProperty, ControlType.MenuItem)` no longer finds all the menu items.

**Comment by Tanya-Solyanik** [comment]
@lovettchris - Is it possible to provide a minimal repro app, with a single control? It seems that regression is between .NET Framework 4.7.2 and .NET Framework 4.8 as well as between .NET7 and .NET 8. Is this correct? What ControlType are you looking for?

**Comment by Zheng-Li01** [comment]
Verified the issue on .NET 7 & .NET 8 & .NET Framework 4.8.1 & 4.7.2, for now  .NET 7.0 & 8.0 &NET Framework 4.8.1 have consistent behavior, which have the different results with .NET Framework 4.7.2, please see below details.

For .NET 7 &  .NET 8, hover over the file menu in the inspect, and notice it has a little **+** sign next to it, but when you click the **+** no sub menu items appear. and after expanding the "File" item, then you can see the subitems in the inspect, such as "New" & "Open"
![image](https://github.com/dotnet/winforms/assets/38325459/cdb42b07-ca0e-4926-9bca-7103b98af3fc)
![Core10109](https://github.com/dotnet/winforms/assets/38325459/d781e21e-2bf3-47bb-997a-fa741f3bdcaa)

For .NET Framework 4.8.1, hover over the file menu in the inspect, and notice it has a little **+** sign next to it, but when you click the **+** no sub menu items appear. and after expanding the "File" item, then you can see the subitems in the inspect, such as "New" & "Open"

For .NET Framework 4.7.2,  hover over the file menu in the inspect, and notice it has a little **+** sign next to it,  when you click the **+** that have sub menu items appear. 
![image](https://github.com/dotnet/winforms/assets/38325459/05656d8c-6305-42fb-bdd2-d79dd73027e6)
![Framework10109](https://github.com/dotnet/winforms/assets/38325459/f87e0e5f-a9b9-4913-a610-bbda029aba3c)

@lovettchris Please check to see if my results are the problem you describe.
@Tanya-Solyanik Is the current behavior of .NET & .NET Fraemwork expected？

I also attach the a minimal repro app, you can use it to verify the issue.
[Issue10109.zip](https://github.com/dotnet/winforms/files/12911519/Issue10109.zip)

**Comment by Tanya-Solyanik** [comment]
Thank you @Zheng-Li01!  My take is that .NET Framework 4.7.2 works correctly because inspect view of the menu matches the visual presentation of the menu. Inspect user has to execute the expand action in order to get to the subitems, same as the application use has to expand the parent menu item in the app. @merriemcgaw  -  what is your take on this issue?

**Comment by lovettchris** [comment]
Thank you @Zheng-Li01! -- yes this is an accurate minimal repro of the problem I'm running into.

**Comment by lovettchris** [comment]
I figured out a work around for now, but many folks may be annoyed that they have to write this workaround, this version tries to be a bit lazy so it doesn't have to open every single menu up front, just the ones that are needed for a given menu item name.

```csharp

        public AutomationWrapper FindMenuItem(string name)
        {
            ReloadMenuItems(name);
            return _menuItems[name];
        }

        void LoadMenuItems(AutomationElement menuItem, List<string> names)
        {
            if (menuItem.TryGetCurrentPattern(ExpandCollapsePattern.Pattern, out object patternObject)
                && patternObject is ExpandCollapsePattern pattern)
            {
                pattern.Expand();
                foreach (AutomationElement subMenuItem in menuItem.FindAll(TreeScope.Descendants,
                    new PropertyCondition(AutomationElement.ControlTypeProperty, ControlType.MenuItem)))
                {
                    var childId = subMenuItem.Current.Name;
                    names.Add(childId);
                    _menuItems[childId] = new AutomationWrapper(subMenuItem);
                    LoadMenuItems(subMenuItem, names);
                }
                pattern.Collapse();
            }
        }

        void ReloadMenuItems(string name)
        {
            if (_menuItems == null)
            {
                _menuItems = new Dictionary<string, AutomationWrapper>();
            }
            if (_menuHeirarchy == null)
            {
                _menuHeirarchy = new Dictionary<string, List<string>>();
            }

            if (_menuItems.Count == 0 || !_menuItems.ContainsKey(name))
            {
                var menuStrip = this._acc.AutomationElement.FindFirst(TreeScope.Children,
                    new PropertyCondition(AutomationElement.ControlTypeProperty, ControlType.MenuBar));

                // enumerate the menu bar and start filling out the _menuHeirarchy.
                // Note: starting in .NET 4.8 one has to literally expand the menu on screen to 
                // find the menu items inside!
                foreach (AutomationElement menuItem in menuStrip.FindAll(TreeScope.Children,
                        new PropertyCondition(AutomationElement.ControlTypeProperty, ControlType.MenuItem)))
                {
                    var id = menuItem.Current.Name;
                    if (!_menuHeirarchy.ContainsKey(id))
                    {
                        List<string> names = new List<string>();
                        LoadMenuItems(menuItem, names);
                        _menuHeirarchy[id] = names;
                        if (names.Contains(name))
                        {
                            // found it!
                            break;
                        }
                    }
                }
            }

            if (!_menuItems.ContainsKey(name))
            {
                throw new Exception($"Could not find menu item {name}");
            }
        }
```

**Comment by lovettchris** [comment]
But now I have a new problem the InvokePattern Invoke() method for my "File/SaveAs" menu item used to return immediately as soon as the open file dialog appears, and then I could automate that dialog to fill in a file name, click Save, etc.  But now, the Invoke method is blocking my unit test entirely until I close the dialog.  This means my unit test cannot automate the save as dialog!!?  I tried putting it in a Task.Run but then I found the next thing in my unittest that tried to do anything with an automation object also blocked?

**Comment by lovettchris** [comment]
Yes, in .NET 4.7.2 the save as dialog does not block the UI Automation SaveAs menu item Invoke method.    In .NET 4.8 (and presumably dotnet 6 and 7 also, although I haven't proved that) it blocks.  But weirdly I have had none of these problems in my [.NET 7.0 WPF application](https://github.com/MoneyTools/MyMoney.Net/tree/master/Source/WPF) which also uses System.Windows.Automation to automate menus and modal dialogs. Could all these problems be specific to WinForms?

**Comment by Tanya-Solyanik** [comment]
@lovettchris  - could you please create a specific issue for Invoke method? This is specific to Winforms.

**Comment by lovettchris** [comment]
@Tanya-Solyanik Done:  https://github.com/dotnet/winforms/issues/10244
